### PR TITLE
#42: Response information was added to the callbak to provide access …

### PR DESCRIPTION
**Pull request by vlavrynovych**
…to the headers

### CODE REVIEW

**Comment by vlavrynovych** [comment]
@UnbounDev please promote it to the master

**Comment by TimoSta** [comment]
This looks like a very valuable addition to me.

However, I would argue that it would make sense to instead pass the whole `response` object to the callback, we should extract the rate limit info and map it to an easily processable format.
This way we A) have a layer of abstraction over the HTTP internals that are not particularly relevant to the caller, and B) the rate info in a format that can immediately be used instead of having to parse it on the side of the caller.

According to the API docs, strava provides the rate info as two HTTP headers with two parts each:
```
X-Ratelimit-Limit: 600,30000
X-Ratelimit-Usage: 642,27300
```

That is comma separated the short term and long term usage and limit.

My suggestion would be to parse those two headers inside the `util._requestHelper` to the following format and pass it to the callback as a third parameter `rateLimit`:

```
{
    "shortTermUsage": 642,
    "shortTermLimit": 600,
    "longTermUsage": 27300,
    "longTermLimit": 30000
}
```

This seems like a cleaner API design to me then just passing the response object back and would still fulfill your requirement of providing rate information.
Your thoughts?

**Comment by vlavrynovych** [comment]
Hi @TimoSta,

Absolutely agree with your format. For me just important to get this information in any representation, because right now I have to calculate it by myself and it's an additional complexity in my code.

Should I prepare a PR according to your description or you can do?

**Comment by TimoSta** [comment]
@vlavrynovych If you don't mind, please go ahead. I will merge it into master then ASAP.

Could you please make sure that there is some unit test coverage of this feature and maybe a short paragraph outlining the feature in the _Usage_ section of the readme? We should strive to improve the quality of the project to keep it maintainable, and I believe proper documentation is an important first step.

Thank you!

**Comment by vlavrynovych** [comment]
Hello @TimoSta,

Could you please help me with tests. I've ran my new test locally and it works perfect.
Also I've tried to run all the tests using `npm test` from the master branch without any changes and they failed. Please assist.

**Comment by vlavrynovych** [comment]
@TimoSta,

`401 Unauthorized` is returning from Strava

File: lib/util.js [hunk 1]
```diff
@@ -211,18 +212,30 @@ util.getRequestBodyObj = function(allowedProps,args) {
...
+function parseRateLimits(headers) {
+    if(!headers[util.rateLimit] || !headers[util.rateUsage]) {
+        return null;
+    }
+
+    var limit = headers[util.rateLimit].split(',')
+        , usage = headers[util.rateUsage].split(',');
+
+    return {
+        shortTermUsage: parseInt(usage[0]),
```
Comment (TimoSta) [inline lib/util.js]: `parseInt()` requires a second parameter `radix`. This may or may not default to the expected value 10 depending on the implementation, but is not guaranteed. Could you please add this?

See https://developer.mozilla.org/en/docs/Web/JavaScript/Reference/Global_Objects/parseInt for details.

**Comment by TimoSta** [comment]
Thanks for making this change.

We had an issue where the test set for the authenticator cleared the `STRAVA_ACCESS_TOKEN` environment variable, which is used by Travis to authenticate the test cases. This causes all following tests to fail due to auth errors. This was resolved with 0ee9444a2646d0445224246707ebac61a2f9b545 on the master.

Additionally, we have a short term limit of 600 calls to the Strava API within 15 minutes. This has also caused tests to fail randomly in the past.

Bottom line is, the server side build againt the actual API is very unreliable and brittle right now. I will write a stub for the Strava API to test against instead when I find some time for it to enable us to reliably run unit tests. Until then, we will have to live with what we have.

I will go ahead and merge this change into the master. In combination with the fix I mentioned above, the test suite should run on the server as well. If not, I'll find out why. :-)

-- Timo

File: lib/util.js (hunk 1 above)
Comment (vlavrynovych) [inline lib/util.js]: Sure

**Comment by TimoSta** [comment]
Looks good to me. Thanks!

**Comment by vlavrynovych** [comment]
@TimoSta,
Thank you too. Finally we can continue work on this cool library.

**Comment by TimoSta** [comment]
+1

The build was successful on the master: https://travis-ci.org/UnbounDev/node-strava-v3/builds/221752993
